### PR TITLE
feat: add 3D LiveChat page

### LINKED ISSUE #879: [casgine] support showing chats in 3D games
See: https://demo-admin.casibase.com/chat

image

Right now the chats are in the chat window. We want to make it in a 3D scene. AI is a human character. The user (we) may be also a human character. It's based on first-person or third-person.

The URL would be: https://demo-admin.casibase.com/livechat

We will use the JS 3D engine: https://threejs.org/

### PR DESCRIPTION
**Pull request by Arcadia-Liu**
Fix: https://github.com/casibase/casibase/issues/879

Initialize 3D scene and human character. Includes five animations: Idle (default), and four talking animations that loop randomly in response to messages.

Rendered the 3D scene using skybox and High Dynamic Range Lighting (HDR). Included a simple water simulation. 

Integrated OrbitControls for camera manipulation: Left mouse button for rotation. Right mouse button for panning. Middle mouse button for scrolling.

UI Enhancement: Added a message list UI component.

External Import Optimization: Import of new dependencies such as threejs via external methods.

TODO: Scenery and animation improvement. External import via casibase-static cdn. UI improvement. Third-person roaming.

Refs: https://github.com/casibase/casibase/issues/879

### CODE REVIEW

**Comment by Sherlocksuper** [comment]
thank you and there are some suggestions for your pr
1. the livechatpage's fetch() is the same as chatpage's fetch(), and livechatpage extends chatpage , so i think the livechatpage's fetch()  can be deleted (? 
2. When I send a message, the livechatbox scroll bar to the top
3. Up to now, Casibase has added some new features, such as suggestions under Message, voice playback, likes, etc. Maybe you can try to join in livechat? (p1、p2)
4. Pleace add i18n for livechat , thanks
5. When I open the livechat interface for the first time, an error is reported （p3）

X)}{QI6`$@J8FO{P}Z{RA}J


~C7$8YW5YLOMTBK2MMA6RWQ


L@{D2Y_KK5785YDJ` _UX5

**Comment by Arcadia-Liu** [comment]
> thank you and there are some suggestions for your pr
> 
> 1. the livechatpage's fetch() is the same as chatpage's fetch(), and livechatpage extends chatpage , so i think the livechatpage's fetch()  can be deleted (?
> 2. When I send a message, the livechatbox scroll bar to the top
> 3. Up to now, Casibase has added some new features, such as suggestions under Message, voice playback, likes, etc. Maybe you can try to join in livechat? (p1、p2)
> 4. Pleace add i18n for livechat , thanks
> 5. When I open the livechat interface for the first time, an error is reported （p3）

Thank you for your suggestions. I've fixed most of them except the third one. I'll continue update all these new features in my future contribution.

File: web/public/index.html [hunk 1]
```diff
@@ -25,6 +25,7 @@
       Learn how to configure a non-root public URL by running `npm run build`.
     -->
     <title>Casibase</title>
+    <script src="https://cdn.jsdelivr.net/npm/three@0.153.0/build/three.min.js"></script>
```
Comment (hsluoyz) [inline web/public/index.html]: jsdelivr is slow, do we have better CDN choices?

File: web/src/App.js [hunk 2]
```diff
@@ -51,13 +51,24 @@ import TaskEditPage from "./TaskEditPage";
...
 import ChatPage from "./ChatPage";
+import LiveChatPage from "./LiveChatPage";
 import CustomGithubCorner from "./CustomGithubCorner";
 import ShortcutsPage from "./basic/ShortcutsPage";
 import UsagePage from "./UsagePage";
 import * as StoreBackend from "./backend/StoreBackend";
 
 const {Header, Footer, Content} = Layout;
 
+// eslint-disable-next-line no-console
```
Comment (hsluoyz) [inline web/src/App.js:62]: What's this?

File: web/src/App.js [hunk 3]
```diff
@@ -309,9 +322,7 @@ class App extends Component {
 
   getMenuItems() {
     const res = [];
-
```
Comment (hsluoyz) [inline web/src/App.js]: Don't change anything unrelared

File: web/src/App.js [hunk 4]
```diff
@@ -469,6 +479,24 @@ class App extends Component {
   }
 
   renderRouter() {
+    const add3DExternals = () => {
+      const scripts = [
+        "https://cdn.jsdelivr.net/npm/three/examples/js/loaders/GLTFLoader.js",
```
Comment (hsluoyz) [inline web/src/App.js:488]: Change CDN

File: web/src/ProvidersUsageTable.js [hunk 5]
```diff
@@ -1,66 +1,66 @@
...
-      }}>
-        <Table rowKey="key" columns={providersUsageColumn} dataSource={dataSource} size="middle" bordered
-          pagination={false} />
-      </div>
-    );
-  }
-}
-
-export default ProvidersUsageTable;
+// Copyright 2024 The casbin Authors. All Rights Reserved.
```
Comment (hsluoyz) [inline web/src/ProvidersUsageTable.js]: Why change this file?

File: web/src/locales/en/data.json [hunk 6]
```diff
@@ -133,7 +133,8 @@
     "Vectors": "Vectors",
     "Videos": "Videos",
     "View": "View",
-    "Wordsets": "Wordsets"
+    "Wordsets": "Wordsets",
+    "Livechat": "Livechat"
```
Comment (hsluoyz) [inline web/src/locales/en/data.json]: i18n is not added manually, just don't mess with i18n in this PR

File: web/yarn.lock [hunk 7]
```diff
@@ -2,34 +2,29 @@
 # yarn lockfile v1
 
 
-"3d-force-graph@1":
-  version "1.71.5"
-  resolved "https://registry.yarnpkg.com/3d-force-graph/-/3d-force-graph-1.71.5.tgz#c36607c242b00b33622dbd118b1512aac7d285db"
-  integrity sha512-h+iLJhvouZz8bmQRxYqoV1OXtG+aJ9LzcakLwR9VUkBJ8tmZx5HtbNm5pYAJ4c1kzZc9JljJzh2u0CCJUrmA4A==
+"3d-force-graph@^1.73":
```
Comment (hsluoyz) [inline web/yarn.lock]: Why lock file changed so much?

File: web/package.json [hunk 8]
```diff
@@ -5,6 +5,7 @@
   "dependencies": {
     "@ant-design/cssinjs": "^1.12.0",
     "@ant-design/icons": "4.6.2",
+    "@babel/preset-react": "^7.24.7",
```
Comment (hsluoyz) [inline web/package.json]: Do you have good reason for this?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: web/yarn.lock (hunk 7 above)
Comment (Arcadia-Liu) [reply web/yarn.lock]: Sorry, I accidentally update the version of @babel/preset-react resulted in a series of sub-dependency updates. Have Fixed now.

File: web/package.json (hunk 8 above)
Comment (Arcadia-Liu) [reply web/package.json]: Same as above. Have fixed now.

File: web/src/locales/en/data.json (hunk 6 above)
Comment (Arcadia-Liu) [reply web/src/locales/en/data.json]: OK, I've deleted my changes in i18n json files.

File: web/src/ProvidersUsageTable.js (hunk 5 above)
Comment (Arcadia-Liu) [reply web/src/ProvidersUsageTable.js]: I can't see any changes in this file. For some reason github detects some changes, but I compared each line carefully and it shows exactly the same changes deleting and adding exactly the same things. I re-downloaded the original file, and my local git didn't detect any changes either.

File: web/public/index.html [hunk 9]
```diff
@@ -25,6 +25,7 @@
       Learn how to configure a non-root public URL by running `npm run build`.
     -->
     <title>Casibase</title>
+    <script src="https://jsd.cdn.zzko.cn/npm/three@0.160.1/build/three.min.js"></script>
```
Comment (hsluoyz) [inline web/public/index.html]: Plz change back to jsdelivr, thx

File: web/src/App.js [hunk 10]
```diff
@@ -469,6 +483,24 @@ class App extends Component {
   }
 
   renderRouter() {
+    const add3DExternals = () => {
+      const scripts = [
+        "https://jsd.cdn.zzko.cn/npm/three/examples/js/loaders/GLTFLoader.js",
```
Comment (hsluoyz) [inline web/src/App.js]: Plz change back to jsdelivr, thx

File: web/src/LiveChatBox.js [hunk 11]
```diff
@@ -0,0 +1,405 @@
+import React from "react";
```
Comment (hsluoyz) [inline web/src/LiveChatBox.js:15]: Apache header

File: web/src/ProvidersUsageTable.js (hunk 5 above)
Comment (hsluoyz) [reply web/src/ProvidersUsageTable.js]: Don't add this file in your commit

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: web/src/App.js [hunk 12]
```diff
@@ -132,10 +143,12 @@ class App extends Component {
       this.setState({selectedMenuKey: "/factorsets"});
     } else if (uri.includes("/videos")) {
       this.setState({selectedMenuKey: "/videos"});
-    } else if (uri.includes("/chat")) {
+    } else if (uri.includes("/chat") && !uri.includes("/livechat")) {
       this.setState({selectedMenuKey: "/chat"});
     } else if (uri.includes("/swagger")) {
       this.setState({selectedMenuKey: "/swagger"});
+    } else if (uri.includes("/livechat")) {
```
Comment (hsluoyz) [inline web/src/App.js]: move up

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Review by hsluoyz — APPROVED** [review]

**Comment by Sherlocksuper** [comment]
> > thank you and there are some suggestions for your pr
> > 
> > 1. the livechatpage's fetch() is the same as chatpage's fetch(), and livechatpage extends chatpage , so i think the livechatpage's fetch()  can be deleted (?
> > 2. When I send a message, the livechatbox scroll bar to the top
> > 3. Up to now, Casibase has added some new features, such as suggestions under Message, voice playback, likes, etc. Maybe you can try to join in livechat? (p1、p2)
> > 4. Pleace add i18n for livechat , thanks
> > 5. When I open the livechat interface for the first time, an error is reported （p3）
> 
> Thank you for your suggestions. I've fixed most of them except the third one. I'll continue update all these new features in my future contribution.

wonderful job i think
but this prompts（eight white box） at p1 is still here,
Did you intentionally leave it there? Perhaps I think it would be better to remove it.

image

**Comment by Arcadia-Liu** [comment]
> > > thank you and there are some suggestions for your pr
> > > 
> > > 1. the livechatpage's fetch() is the same as chatpage's fetch(), and livechatpage extends chatpage , so i think the livechatpage's fetch()  can be deleted (?
> > > 2. When I send a message, the livechatbox scroll bar to the top
> > > 3. Up to now, Casibase has added some new features, such as suggestions under Message, voice playback, likes, etc. Maybe you can try to join in livechat? (p1、p2)
> > > 4. Pleace add i18n for livechat , thanks
> > > 5. When I open the livechat interface for the first time, an error is reported （p3）
> > 
> > 
> > Thank you for your suggestions. I've fixed most of them except the third one. I'll continue update all these new features in my future contribution.
> 
> wonderful job i think but this prompts（eight white box） at p1 is still here, Did you intentionally leave it there? Perhaps I think it would be better to remove it.

Hi, thank you for pointing this out. Could you please be more specific as to the conditions under which these white boxes appear? I've pulled the latest version. These suggestion prompts can be displayed correctly on chatpage. However, I can't see these contents show up in my livechat page.
BTW: New commits are to resolve compatibility issues caused by the latest merged changes, ensuring proper functionality of the LiveChat page. No new features have been added.
image
image

**Comment by Sherlocksuper** [comment]
> > > > thank you and there are some suggestions for your pr
> > > > 
> > > > 1. the livechatpage's fetch() is the same as chatpage's fetch(), and livechatpage extends chatpage , so i think the livechatpage's fetch()  can be deleted (?
> > > > 2. When I send a message, the livechatbox scroll bar to the top
> > > > 3. Up to now, Casibase has added some new features, such as suggestions under Message, voice playback, likes, etc. Maybe you can try to join in livechat? (p1、p2)
> > > > 4. Pleace add i18n for livechat , thanks
> > > > 5. When I open the livechat interface for the first time, an error is reported （p3）
> > > 
> > > 
> > > Thank you for your suggestions. I've fixed most of them except the third one. I'll continue update all these new features in my future contribution.
> > 
> > 
> > wonderful job i think but this prompts（eight white box） at p1 is still here, Did you intentionally leave it there? Perhaps I think it would be better to remove it.
> 
> Hi, thank you for pointing this out. Could you please be more specific as to the conditions under which these white boxes appear? I've pulled the latest version. These suggestion prompts can be displayed correctly on chatpage. However, I can't see these contents show up in my livechat page. BTW: New commits are to resolve compatibility issues caused by the latest merged changes, ensuring proper functionality of the LiveChat page. No new features have been added. image image

ok, my mistake

**Pull request merged**